### ISSUE TITLE
Write rake task to bulk fetch OCR text given a list of barcodes

### ISSUE THREAD

**Issue by peetucket** [sul-dlss/google-books#146]
One of the primary use cases of accessioned Google Books is bulk access to the OCRed content.  This use case is not well served by anything in our current infrastructure, but would be served by a future SDR access API.  In the interim, we can design a way to provide content to key stakeholders.

We have two specific researchers to serve for this use case, with ~75K barcodes in each use case (see #327 for the files with barcodes).

The proposed system would assume:

1. A user has provided a list of barcodes.
1. A mount point or disk location we could move files to.
1. The user will not remove files from the mount point before all books have been transferred (i.e. we can rely on the presence of files in the destination mount point as confirmation the book has been transferred successfully)

The system should be able to:

1. Determine if a barcode has been accessioned.
1. If not accessioned, skip it.
1. If accessioned, move any OCR files (.txt., .xml., .html) for that book to specified mount point in a folder based on barcode.
1. Log activity in some form needed for debugging/status viewing.
1. Know when it has copied all files for a book.
1. Restart if needed and be able to skip over books it already copied files for.

The system will be designed to be run by an engineer and can exist on server under our control that has ready only access to the preservation mounts and write access to the mounts needed for copying content to.

* script should take as inputs: (a) a text file with barcodes of interest, (b) a specifying files to extract from object, (c) a destination path
* script has to lookup barcodes to get druids
* script has to use preservation catalog API to determine disk where content is found
* script can use the presence of the output file/folder as confirmation that the book has been transferred completely, and therefore does not need state persistence in a database
* script can verify if something is accessioned into preservation by querying for checksums from preservation using the checksum report API endpoint

Uncertainty: we still don't know exactly where the content will be staged.  Could assume a mount for now.

**Comment by jcoyne** [comment]
@peetucket who will run this script?  Who is the person who needs this content?   Do we have to enforce copyright restrictions on this data? How often will we have to do this?

**Comment by peetucket** [comment]
> who will run this script? 

a developer or someone who has access to the server and has been trained

> Who is the person who needs this content? 

designated faculty that are working with SUL 

> Do we have to enforce copyright restrictions on this data? 

no, that would be up to the faculty member to ensure compliance

> How often will we have to do this?

guesstimate: once a quarterish

**Comment by jcoyne** [comment]
>>    who will run this script?

> a developer or someone who has access to the server and has been trained

>> How often will we have to do this?

> guesstimate: once a quarterish

Have we budgeted for that time?  Are we going to do it into perpetuity? Do we have a manager who agrees to that?
>>   Who is the person who needs this content?
> designated faculty that are working with SUL



>>  Do we have to enforce copyright restrictions on this data?

> no, that would be up to the faculty member to ensure compliance

Need a written agreement on that?


How will we identify which items they need access to?  Since this workcycle isn't about getting all of the data harvested, can we move this job into it's own workcycle?

**Comment by peetucket** [comment]
> Have we budgeted for that time? Are we going to do it into perpetuity? Do we have a manager who agrees to that?

All good questions. This was a compromise proposal I made as a way to provide some user benefit to our stakeholders.  Its still less effort that any more appealing alternatives I can think of (like building a proper SDR access API).

> Need a written agreement on that?
> 

Indeed, but this is not in our responsibility.

> How will we identify which items they need access to? Since this workcycle isn't about getting all of the data harvested, can we move this job into it's own workcycle?

Possibly, though one of the key stakeholders indicated they would be satisfied with getting data as it arrives.

**Comment by peetucket** [comment]
If we provide content incrementally: How will be know what we have given and what they already have?

**Comment by ndushay** [comment]
Imagine a world where a requestor could paste in the searchworks URL for the results of interest and we could work from that ...

**Comment by peetucket** [comment]
"Alexa, give me all of the books about law from 1960s onward and put them in my Dropbox."

**Comment by ndushay** [comment]
in all seriousness, a SearchWorks URL is possible, translating that into a solr query to sw solr is possible, and getting the barcodes and ckeys from the solr response is possible.  But that's really for the other ticket.

**Comment by peetucket** [comment]
Yes, that's possible - though it assumes the metadata a researcher would use for filtering is all available in Searchworks.  Though if we were doing a more general case system from start to finish, that would be a good starting approach.  For this particular case, we can assume its been done and we are starting with barcodes.

**Comment by cncoleman** [comment]
Receiving content incrementally is not preferred, it was offered as a compromise if: 1. Researcher could not request specific content (submit a list of barcodes) and 2. The researcher would otherwise have to wait 3+ years for all of the content to be in SDR.  If a researcher requests a specific subset of data, like, everything with LCSH "LAW" it is preferable for them to get it all at once. They will not work on it piecemeal, but will take the entire corpus, analyze it, understand the shape of it, decide what tasks they want to use it for.  A log that lets them know how many requested barcodes did not return items would be very helpful.

**Comment by peetucket** [comment]
Ok, thanks.  Given that for each use case, we identified ~75K barcodes, this means we are looking at 150k barcodes total.  We can definitely prioritize those to the top, so IF we were able to turn the knob on in production in the next month or two (still an IF based on some other factors), it is possible that all of those books would be done within this calendar year (assuming ~1000-1500 books per day, which we've proven is possible in -stage in a couple of days of testing).  So I presume this would allow us to deliver things all at once instead of incrementally?  This may not substantially change the proposed system design we have discussed so far, but is good to know.
